onnxruntime/core/framework/resource_accountant.cc

```diff
@@ -0,0 +1,188 @@
+#include "core/common/safeint.h"
+#include "core/common/string_utils.h"
+
+#include "core/framework/config_options.h"
+#include "core/graph/constants.h"
+#include "core/session/onnxruntime_session_options_config_keys.h"
+
+#include <fstream>
+
+namespace onnxruntime {
+
+// Use this accountant if your resource can be counted with size_t type
+class SizeTAccountant : public IResourceAccountant {
+ public:
+  SizeTAccountant() = default;
+  ~SizeTAccountant() = default;
+
+  SizeTAccountant(size_t threshold, InlinedHashMap<std::string, NodeAllocationStats>&& node_stats)
+      : IResourceAccountant(threshold), node_stats_(std::move(node_stats)) {}
+
+  explicit SizeTAccountant(InlinedHashMap<std::string, NodeAllocationStats>&& node_stats)
+      : IResourceAccountant(), node_stats_(std::move(node_stats)) {}
+
+  ResourceCount GetConsumedAmount() const noexcept override {
+    return consumed_amount_;
+  }
+
+  void AddConsumedAmount(const ResourceCount& amount) noexcept override {
+    if (std::holds_alternative<size_t>(amount)) {
+      consumed_amount_ += std::get<size_t>(amount);
+    }
+  }
+  void RemoveConsumedAmount(const ResourceCount& amount) noexcept override {
+    if (std::holds_alternative<size_t>(amount)) {
+      consumed_amount_ -= std::get<0>(amount);
+    }
+  }
+
+  ResourceCount ComputeResourceCount(const std::string& node_name) const override {
+    auto hit = node_stats_.find(node_name);
+    if (hit != node_stats_.end()) {
+      const auto& stats = hit->second;
+      return stats.input_sizes + stats.initializers_sizes +
+             stats.total_dynamic_sizes + stats.total_temp_allocations;
+    }
+    return static_cast<size_t>(0U);
+  }
+
+ private:
+  size_t consumed_amount_ = 0;
+  InlinedHashMap<std::string, NodeAllocationStats> node_stats_;
+};
+
+struct NodeStatsRecorder::Impl {
+  std::filesystem::path node_stats_path;
+  // This is a node name to allocation stats map
+  InlinedHashMap<std::string, NodeAllocationStats> node_stats;
+  // Keeps track of nodes for which input/output sizes are accounted
+  InlinedHashSet<std::string> input_output_accounted;
+};
+
+NodeStatsRecorder::NodeStatsRecorder(const std::filesystem::path& node_stats_path)
+    : impl_(std::make_unique<Impl>()) {
+  impl_->node_stats_path = node_stats_path;
+}
+
+NodeStatsRecorder::~NodeStatsRecorder() = default;
+
+const std::filesystem::path& NodeStatsRecorder::GetNodeStatsFileName() const noexcept {
+  return impl_->node_stats_path;
+}
+
+bool NodeStatsRecorder::ShouldAccountFor(const std::string& input_output_name) const {
+  return impl_->input_output_accounted.insert(input_output_name).second;
+}
+
+void NodeStatsRecorder::ResetPerRunNameDeduper() {
+  impl_->input_output_accounted.clear();
+}
+
+void NodeStatsRecorder::ReportNodeStats(const std::string& node_name, const NodeAllocationStats& stats) {
+  auto result = impl_->node_stats.emplace(node_name, stats);
+  if (!result.second) {
+    // Node already exists, update the stats
```

Comment:
I think each node is visited only once (only assigned to one EP). In what situation will the code hit this?

Reply:
This to allow multiple runs. Can the memory usage grow for attention kernel?

Reply:
That makes sense. Previously I though we only record one inference run. 
Adding some comments can help reader.